### PR TITLE
[WIP] Feat/gpt oss + peagle example

### PR DESCRIPTION
**Pull request by ymwangg**
Add p-eagle gpt-oss-20b example.

### CODE REVIEW

**Comment by ymwangg** [comment]
# gpt-oss + P-EAGLE — Implementation Status Report

_Status as of the current branch `feat/gpt-oss-example`. Target: gpt-oss-20b on AWS
Trainium (TP=4) via NKIPy._

## TL;DR

- **Base gpt-oss-20b**: complete and working on Trainium (~70 tok/s decode).
- **P-EAGLE speculative decoding**: numerically correct and fully on-device. The
  drafter port (the recent work) is done and validated — device drafter is
  byte-identical to the GPU-validated CPU reference and **~6× faster than it**.
- **Open issue**: P-EAGLE end-to-end is currently a **net slowdown vs the base
  decode (~0.57×)** despite accepting ~3.4 tokens/step, because the speculative
  loop is synchronous while the base decode is async/double-buffered. The drafter
  is no longer the bottleneck — the loop is.

## Components

| Component | File(s) | Status |
|---|---|---|
| Base MoE model (TP, MXFP4→bf16, sinks, sliding window, YaRN RoPE, double-buffered decode) | `gpt_oss.py`, `kernels/`, `tensor_preparation.py` | ✅ Working on hardware |
| Drafter — CPU reference (KV-cached full-context parallel drafting) | `eagle/drafter_cpu.py` | ✅ GPU-validated vs vLLM |
| Drafter — on-device (KV-cached forward + draft head) | `eagle/drafter_model.py`, `eagle/kernels/drafter*.py` | ✅ Ported + Trainium-validated |
| Speculation loop (prefill → draft → verify → accept) | `eagle/speculate.py` | ✅ Working; `--device-drafter` selects on-device path |
| Verify kernel (K+1 multi-token target pass + per-pos argmax) | `eagle/kernels/verify.py` | ✅ Working |
| Tests | `eagle/test_drafter_cpu.py`, `eagle/test_drafter_device.py` | ✅ Pass (real + synthetic) |

## What changed recently (this work)

The drafter had a working CPU reference but the on-device path was the **old
context-blind kernel** (K positions under a static `(K,K)` cross-depth mask, no
prefill, no KV cache) and was unused. Three commits brought the device path up to
parity and tightened metrics:

| Commit | Summary |
|---|---|
| `fdb13e0` | fix: correct P-EAGLE drafter to full-context KV-cached parallel drafting (CPU reference + loop bookkeeping; GPU-validated vs vLLM) |
| `e8da8b1` | feat: port P-EAGLE drafter to on-device KV-cached parallel drafting (`DrafterModel` + `kernels/drafter*.py`; `--device-drafter`; `test_drafter_device.py`) |
| `2cd9e2f` | fix: correct acceptance-length metric to count only emitted tokens (excludes the final step truncated by the token budget) |

Port design note: the device drafter keeps **static kernel shapes** — it commits
newly-accepted tokens one at a time (`S=1`) and runs the K-wide draft window
(`S=K`) — so it avoids per-step recompiles. Causal attention makes the
one-at-a-time commits numerically identical to a single batched forward.

Also fixed along the way: `speculate.py`'s chat-template branch crashed on recent
transformers (it returns a `BatchEncoding`, not an array); now extracts
`["input_ids"]`. The template ships separately as `chat_template.jinja`.

## Validation

| Check | Result |
|---|---|
| CPU drafter math vs vLLM (GPU) | cos 0.9999 prefill; 100% draft-token match |
| Device drafter vs CPU drafter (synthetic, Trainium) | logit cos 0.9999 across prefill + KV-cached draft steps with rollback |
| Device vs CPU drafter, end-to-end (real checkpoints) | **byte-identical output** |
| `test_drafter_cpu.py` (real checkpoint) | 3/3 pass |
| `test_drafter_device.py` (Neuron-gated) | pass |

Under **greedy** verification emitted tokens are the target's argmax regardless of
the drafter, so identical output is expected by construction — the drafter only
affects speed. Acceptance length is the meaningful correctness signal (a broken /
context-blind drafter collapses toward ~1); we observe ~2.4–4.6.

## Performance (5-prompt sweep, n=128, K=7, TP=4, chat template, identical env)

| Config | Mean decode tok/s | vs base | Mean acceptance |
|---|---:|---:|---:|
| Base (no speculation) | **69.8** | 1.00× | — |
| P-EAGLE, device drafter | 39.8 | **0.57×** | 3.44 |
| P-EAGLE, CPU drafter | 6.6 | 0.10× | 3.37 |

- Device drafter is **~6× faster than the CPU drafter** (the port's contribution).
- Speedup tracks acceptance: 0.72× on high-acceptance prompts (haiku 4.59,
  binary-search 4.38) down to 0.44× on low-acceptance ones (transformer 2.44).
- None exceed 1× — see the bottleneck below.

(Full per-prompt table in `gpt-oss-peagle-results.md`.)

## Known issues / open work

1. **Speculation is a net slowdown vs base decode (primary issue).** The base path
   is async / double-buffered (overlapped execution); the speculative loop is
   synchronous, paying every step for a K+1-wide verify pass plus host-side
   orchestration (accept/reject compare, embedding lookups, drafter call) that is
   not pipelined. To make P-EAGLE a net win:
   - async / double-buffered speculation (overlap verify + draft + sampling);
   - move per-step host orchestration on-device to cut host round-trips;
   - tune K per workload (lower K helps low-acceptance prompts).

2. **Acceptance below model card on some prompts.** Chat formatting lifts
   acceptance (raw prompts ≈2.95 → chat 4.1–4.6 on favorable prompts), but
   reasoning-heavy prompts still sit ~2.4–2.5. Worth checking tap-layer choice and
   prompt distribution against the card's eval setup.

3. **Single-run benchmarks.** Numbers are one run per (prompt, config) — good for
   relative comparison; average several runs before quoting as benchmarks.

**Comment by ymwangg** [comment]
## Progress update: P-EAGLE speculative decoding — working + optimized on trn2

Brought the on-device P-EAGLE drafter from **broken → 62 tok/s** (vs 52 tok/s for the base model), and profiled the verify step to a firm conclusion. All changes are on this branch (4 commits).

### What changed

| Commit | Change |
|---|---|
| `fcdb576b` | **KV-cached on-device drafter (default).** The device drafter was context-blind (no KV cache/prefill) → acceptance collapsed to ~1.5. Ported the CPU drafter's full-context KV-cache algorithm to device. Also fixed an `apply_chat_template` crash on transformers 5.5.4. |
| `5fe2a70f` | **Fold commit into the draft pass.** Replaced 4 sequential width-1 "commit" kernel launches/step with one combined forward of width `W=C+K-1`. |
| `e505269b` | **Fuse the drafter into one kernel.** fc-fusion + `[embed\|hidden]` assembly + all 4 layers (each aliasing its KV cache) + norm + lm_head in a single device call, replacing the per-layer host loop. |
| `a5d15b09` | **Default K=3** (throughput optimum) + documented verify scaling. |

Parity held throughout: device drafter matches the CPU reference at logit cosine **0.9999** (token diffs are sub-noise bf16 ties). CPU drafter tests 3/3.

### Throughput progression (trn2, TP=4, K=7, chat prompt)

| Stage | Acceptance | Decode tok/s |
|---|---|---|
| Base model (no speculation) | — | 52.4 |
| CPU drafter (reference) | 4.33 | 4.8 |
| Device drafter, context-blind | 1.56 | 20.5 |
| + KV cache | 4.33 | 34.5 |
| + commit folded into draft | 4.33 | 40.9 |
| + fused single kernel | 4.33 | 50.4 |
| **+ K=3 (optimum)** | **3.31** | **~62** |

### Latency breakdown (per verify step, SPEC_PROFILE=1)

At the fused/optimized state, K=7 (~116 ms/step):

```
draft :  17.4 ms/step (21%)   1 fused kernel
verify:  66.1 ms/step (78%)   24 MoE layers over K+1 tokens
  layers:  63.4 ms   (aux copies 0.3)
  argmax:   2.3 ms   (head + argmax + host transfer)
other :   0.4 ms/step (0%)
```

Reference points:
- **Single-token base forward: ~19 ms** (embed → 24 MoE layers → norm → lm_head → sample).
- **Verify scales as ~19 ms fixed + 6.3 ms/candidate token.** The fixed ~19 ms == a single-token forward (weight streaming from HBM), amortized across the K+1 batch.

### Choosing K — why 3, not 7

Verify cost grows ~6.3 ms/token, but this drafter's acceptance **plateaus at ~3.3**, so K>3 buys verify time for no extra accepted tokens. Sweep (n=96):

| K | Acceptance | Decode tok/s |
|---|---|---|
| 1 | 1.94 | 50.0 |
| 2 | 2.67 | 52.6 |
| **3** | **3.31** | **~60–62** |
| 4 | 3.31 | 49.5 |
| 7 | 4.33 | 50.4 |

### Can verify be sped up with a better kernel? Investigated → no (for this framework)

Verify is 78% of each step, so I profiled the MoE hard:
- **Ruled out** attention-over-full-KV-cache (max_seq_len 4096→512 barely moved verify: 36 vs 38 ms) and HBM bandwidth (implied ~80 GB/s, well below peak).
- **Cost model:** a single expert FFN is *flat* in row count (0.80 ms at S=1/4/8) → MoE is **weight-load bound**; layer cost ≈ (#expert-weight-loads) × ~1.4 ms.
- The per-token loop does `S×top_k` loads (K=3→16, K=7→32); the batch only touches ~9.7 (K=3) / 13.4 (K=7) **unique** experts. Grouped-all-experts (32 loads, flat) only wins when `S×top_k > 32` → **ties at K=7, regresses at K=3**. Deduping to unique experts *would* win but needs a **dynamic** expert loop the static-trace NKI kernel can't express.

**Conclusion:** verify is near-optimal on this framework — no worthwhile *static* MoE-kernel win. Grouped MoE would help large-L **prefill/TTFT** (`L×top_k`≈280 → 32 loads) if revisited.

### Remaining levers (algorithmic, not kernel)
- **Stronger drafter** to raise the ~3.3 acceptance plateau — would let larger K pay off again.
- **TTFT:** the first `prefill()` compiles a full-prompt-width fused kernel inside the timed region; pre-compiling bucketed prompt widths would hide it.

Profiling is behind `SPEC_PROFILE=1` (zero overhead when unset).

**Comment by ymwangg** [comment]
## Follow-up: verify latency breakdown — where the ~63 ms goes

Attributed the verify layer's cost to its components by microbenchmarking each at the real gpt-oss-20b sharded dims (TP=4, S=K+1=8 tokens), × 24 layers:

| Component | ms/layer | × 24 layers | share |
|---|---|---|---|
| Attention projections (QKV + O) | 0.28 | 6.8 | ~10% |
| MoE router (matmul + topk + softmax) | 0.23 | 5.5 | ~8% |
| **MoE experts (top_k FFNs)** | **5.96** | **143** | **~92%** |
| Attention scores / softmax | — | ~0 | ruled out (seq-len-insensitive) |

*(Isolated components sum higher than the fused 63 ms because fusion overlaps DMA with compute and shares intermediates — the proportions are the point.)*

**Which kernel to focus on: the MoE expert FFN loop, and only that.** Attention (~10%) and the router (~8%) aren't worth touching.

### Important caveat

The per-expert matmul is *not* slow — it's already efficient. The cost is **structural**: the verify batch does `S×top_k` separate expert-weight loads (32 at K=7, ~12.4 MB each), and the MoE is weight-load bound (a single expert FFN is flat in row count: 0.80 ms at S=1/4/8). The only lever is loading **fewer distinct experts**:

- **Dedup** to the ~13 unique experts the batch actually touches → the real win, but needs a **dynamic** expert loop the static-trace NKI kernel can't express (loop counts/shapes must be compile-time constants).
- **Grouped-all-experts** (32 loads, flat) → ties at K=7, regresses at K=3. Not a win in the useful range.

### So, to keep improving verify

1. **Framework-level — dynamic expert dispatch** (gather/scatter over a variable unique-expert set). This is the one change that would actually cut the ~143 ms, but it's an NKIPy/compiler capability, not an example-level kernel edit.
2. **Algorithmic — a stronger drafter** to raise the ~3.3 acceptance plateau, so the fixed per-step cost amortizes over more accepted tokens (and larger K pays off again).

Net: verify is near-optimal for what this static-trace framework can express today; the remaining wins are framework or drafter improvements, not a rewrite of the expert kernel.
